staging/src/k8s.io/client-go/resolver/resolver.go

```diff
@@ -0,0 +1,349 @@
+	if err != nil {
+		return nil, err
+	}
+
+	if netutils.ParseIPSloppy(host) != nil {
+		return net.DefaultResolver, fmt.Errorf("APIServerResolver only works for domain names")
+	}
+
+	// defaulting
+	r := &resolver{
+		host:    host,
+		port:    port,
+		timeout: 100 * time.Second,
+		refresh: &atomicBool{1},
+	}
+
+	// options
+	for _, o := range options {
+		o(r)
+	}
+
+	f := &dns.MemResolver{
+		LookupIP: func(ctx context.Context, network, host string) ([]net.IP, error) {
+			return r.lookupIP(ctx, network, host)
+		},
+	}
+
+	resolver := dns.NewMemoryResolver(f)
+	config.Resolver = resolver
+
+	client, err := rest.RESTClientFor(&config)
+	if err != nil {
+		return nil, err
+	}
+
+	r.client = client
+	// Initialize cache and close idle connections so next connections goes
+	// directly to one of the API server IPs, this is useful for cases
+	// that use a load balancer for bootstrapping
+	r.refreshCache(ctx)
+	utilnet.CloseIdleConnectionsFor(r.client.Client.Transport)
+	return resolver, nil
+}
+
+func setConfigDefaults(config *rest.Config) error {
+	gv := v1.SchemeGroupVersion
+	config.GroupVersion = &gv
+	config.APIPath = "/api"
+	config.NegotiatedSerializer = scheme.Codecs.WithoutConversion()
+
+	if config.UserAgent == "" {
+		config.UserAgent = rest.DefaultKubernetesUserAgent()
+	}
+
+	return nil
+}
+
+func (r *resolver) lookupIP(ctx context.Context, network, host string) ([]net.IP, error) {
+	// Use the default resolver if is not trying to resolve the configured API server hostname
+	if !strings.HasPrefix(host, r.host) {
+		klog.V(7).Infof("Resolver Trace: use default resolver for host %s, different than API server hostname %s", host, r.host)
+		return net.DefaultResolver.LookupIP(ctx, network, host)
+	}
+
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	// Use the default resolver if the cache is empty and try to renew the cache.
+	if len(r.cache) == 0 {
+		klog.V(7).Infof("Resolver Trace: use default resolver for host %s, cache is empty", host)
+		addrs, err := net.DefaultResolver.LookupIP(ctx, network, host)
+		if err != nil {
+			return addrs, err
+		}
+		// don't try to renew the cache until we know it's possible
+		if len(addrs) > 0 {
+			klog.V(7).Infof("Resolver Trace: refreshing cache for first time for host %s", host)
+			go r.refreshCache(ctx)
+		}
+		return addrs, nil
+	}
+
+	// Check if is the first time we query the cache and record the time, it is a fresh cache
+	// so don't try to refresh it again, if something goes wrong next time will be refreshed.
+	// If is not it the first time, it means that previous connection didn't work out
+	// so we return the IPs in a different order to reach a different API server.
+	if r.timestamp.IsZero() {
+		r.timestamp = time.Now()
+	} else {
+		rand.Shuffle(len(r.cache), func(i, j int) {
+			r.cache[i], r.cache[j] = r.cache[j], r.cache[i]
+		})
+		if r.refresh.get() {
+			klog.V(7).Infof("Resolver Trace: refreshing cache for host %s", host)
+			go r.refreshCache(ctx)
```

Comment:
does this mean we reuse an old cache? I don't think we can do that.

Reply:
you have to reuse the cache until you are able to reconnect, you are reusing the client.
In other words, the only chance to refresh the cache is that one of the IPs there works , because if you are here means that you are not able to connect.

Reply:
It's not enough that an IP works. We must not connect to an unready apiserver. By reusing the cache this might happen. I still think we have to do this totally synchronous, potentially with a second client.

Reply:
that is handled by the logic of the client, it retries until some IP works, and works means is able to get the endpoints

Reply:
a working IP is not enough. You will get a response for unready apiservers. But the answer might be wrong.